src/Interpreters/AsynchronousInsertQueue.cpp

```diff
@@ -276,10 +284,9 @@ void AsynchronousInsertQueue::busyCheck()
 {
     auto timeout = busy_timeout;
 
-    while (!shutdown)
+    std::unique_lock shutdown_lock(shutdown_mutex);
+    while (!shutdown_cv.wait_for(shutdown_lock, timeout, [this]() { return shutdown; }))
```

Comment:
It's also weird that condvar is used to sleep, not to wait for some event

Reply:
Actually I think it is fine, since it not only sleep, but also wait for shutdown.
There are other ways of doing so (pipes, timerfd, ...), but anyway it requires kernel interaction, or busy poll (and here it is not acceptable).
And it is used in other places for the same purpose

P.S. Note, that Poco::Event has mutex+condvar inside.